src/main/java/build/buf/protovalidate/internal/evaluator/EvaluatorBuilder.java

```diff
@@ -78,358 +78,404 @@ public EvaluatorBuilder(Env env, boolean disableLazy) {
     }
+    return build(desc);
   }
 
   /**
    * Either returns a memoized {@link Evaluator} for the given descriptor, or lazily constructs a
    * new one.
    */
   private Evaluator build(Descriptor desc) throws CompilationException {
-    Evaluator eval = evaluatorMap.get(desc);
+    Evaluator eval = evaluatorCache.get(desc);
     if (eval != null) {
       return eval;
     }
-    MessageEvaluator msgEval = new MessageEvaluator();
-    evaluatorMap.put(desc, msgEval);
-    buildMessage(desc, msgEval);
-    return msgEval;
-  }
-
-  private void buildMessage(Descriptor desc, MessageEvaluator msgEval) throws CompilationException {
-    try {
-      DynamicMessage defaultInstance =
-          DynamicMessage.newBuilder(desc).mergeFrom(new byte[0], EXTENSION_REGISTRY).buildPartial();
-      Descriptor descriptor = defaultInstance.getDescriptorForType();
-      MessageConstraints msgConstraints =
-          resolver.resolveMessageConstraints(descriptor, EXTENSION_REGISTRY);
-      if (msgConstraints.getDisabled()) {
-        return;
+    synchronized (this) {
+      // Check again (we may have lost race with another thread which populated the map with this
+      // descriptor).
+      eval = evaluatorCache.get(desc);
+      if (eval != null) {
+        return eval;
+      }
+      // Rebuild cache with this descriptor (and any of its dependencies).
+      ImmutableMap<Descriptor, Evaluator> updatedCache =
+          new DescriptorCacheBuilder(env, constraints, evaluatorCache).build(desc);
+      evaluatorCache = updatedCache;
+      eval = updatedCache.get(desc);
+      if (eval == null) {
+        throw new IllegalStateException(
+            "updated cache missing evaluator for descriptor - should not happen");
       }
-      processMessageExpressions(descriptor, msgConstraints, msgEval, defaultInstance);
-      processOneofConstraints(descriptor, msgEval);
-      processFields(descriptor, msgEval);
-    } catch (InvalidProtocolBufferException e) {
-      throw new CompilationException("failed to parse proto definition: " + desc.getFullName(), e);
     }
+    return eval;
   }
 
-  private void processMessageExpressions(
-      Descriptor desc,
-      MessageConstraints msgConstraints,
-      MessageEvaluator msgEval,
-      DynamicMessage message)
-      throws CompilationException {
-    List<Constraint> celList = msgConstraints.getCelList();
-    if (celList.isEmpty()) {
-      return;
-    }
-    Env finalEnv =
-        env.extend(
-            EnvOption.types(message),
-            EnvOption.declarations(
-                Decls.newVar(Variable.THIS_NAME, Decls.newObjectType(desc.getFullName()))));
-    List<CompiledProgram> compiledPrograms = compileConstraints(celList, finalEnv);
-    if (compiledPrograms.isEmpty()) {
-      throw new CompilationException("compile returned null");
-    }
-    msgEval.append(new CelPrograms(compiledPrograms));
-  }
+  private static class DescriptorCacheBuilder {
```

Comment:
I'd recommend viewing the rest of this ignoring whitespace - it is really just moving functions inside this inner static class.